test/azure_blobs_exception_tests.jl

```diff
@@ -144,3 +144,141 @@
+    end
+
+    @testset "403: Forbidden" begin
+        # Returned when you pass an invalid api-key.
+        nretries = test_status(:GET, 403)
+        @test nretries == 1 broken=true
+        nretries = test_status(:PUT, 403)
+        @test nretries == 1 broken=true
+    end
+
+    @testset "404: Not Found" begin
+        nretries = test_status(:GET, 404)
+        @test nretries == 1
+    end
+
+    @testset "405: Method Not Supported" begin
+        nretries = test_status(:GET, 405, ["Allow" => "PUT"])
+        @test nretries == 1 broken=true
+        nretries = test_status(:PUT, 405, ["Allow" => "GET"])
+        @test nretries == 1 broken=true
+    end
+
+    @testset "409: Conflict" begin
+        # Returned when write operations conflict.
+        # NOTE: We currently don't retry but maybe we should? This is probably a case where the
+        # retry logic should add more noise to the backoff so that multiple writers don't collide on retry.
+        nretries = test_status(:GET, 409)
+        @test nretries == max_retries broken=true
+        nretries = test_status(:PUT, 409)
+        @test nretries == max_retries broken=true
+    end
+
+    @testset "412: Precondition Failed" begin
+        # Returned when an If-Match or If-None-Match header's condition evaluates to false
+        nretries = test_status(:GET, 412)
+        @test nretries == 1
+        nretries = test_status(:PUT, 412)
+        @test nretries == 1
+    end
+
+    @testset "413: Content Too Large" begin
+        # https://learn.microsoft.com/en-us/rest/api/storageservices/put-blob?tabs=shared-access-signatures#remarks
+        nretries = test_status(:PUT, 413)
+        @test nretries == 1 broken=true
+    end
+
+    @testset "429: Too Many Requests" begin
+        # TODO: We probably should respect the Retry-After header, but we currently don't
+        # (and we don't know if Azure actually sets it)
+        # NOTE: This can happen when Azure is throttling us, so it might be a good idea to retry with some
+        # larger initial backoff (very eager retries probably only make the situation worse).
+        nretries = test_status(:GET, 429, ["Retry-After" => 10])
+        @test nretries == max_retries broken=true
+        nretries = test_status(:PUT, 429, ["Retry-After" => 10])
+        @test nretries == max_retries broken=true
+    end
+
+    @testset "502: Bad Gateway" begin
+        # This error occurs when you enter HTTP instead of HTTPS in the connection.
+        nretries = test_status(:GET, 502)
+        @test nretries == 1 broken=true
+        nretries = test_status(:PUT, 502)
+        @test nretries == 1 broken=true
+    end
+
+    @testset "503: Service Unavailable" begin
+        # NOTE: This seems similar to 429 and the Azure docs specifically say:
+        #    Important: In this case, we highly recommend that your client code back off and wait before retrying
+        nretries = test_status(:GET, 503)
+        @test nretries == max_retries broken=true
```

Comment:
I think the only problem with this test now is an off-by-one error, where maybe ntries or max_retries is only counting real retries while the other could be counting the first request.